sycl/doc/extensions/LevelZeroBackend/LevelZeroBackend.md

```diff
@@ -58,39 +62,170 @@ and they must be included in the order shown:
-|``` make<event>(const context &, ze_event_handle_t, ownership = transfer);```| Constructs a SYCL event instance from a Level-Zero ```ze_event_handle_t```. The context argument must be a valid SYCL context encapsulating a Level-Zero context. The Level-Zero event should be allocated from an event pool created in the same context. The ```ownership``` argument specifies if the SYCL runtime should take ownership of the passed native handle. The default behavior is to transfer the ownership to the SYCL runtime. See section 4.4 for details.|
-|``` make<program>(const context &, ze_module_handle_t);```| Constructs a SYCL program instance from a Level-Zero ```ze_module_handle_t```. The context argument must be a valid SYCL context encapsulating a Level-Zero context. The Level-Zero module must be fully linked (i.e. not require further linking through [```zeModuleDynamicLink```](https://spec.oneapi.com/level-zero/latest/core/api.html?highlight=zemoduledynamiclink#_CPPv419zeModuleDynamicLink8uint32_tP18ze_module_handle_tP28ze_module_build_log_handle_t)), and thus the SYCL program is created in the "linked" state.|
+The following free functions defined in the ```sycl``` namespace are specialized for Level-Zero backend to allow
+an application to create a SYCL object that encapsulates a corresponding Level-Zero object:
+
+<table>
+<tr>
+<th>Level-Zero interoperability function</th>
+<th style="text-align:left"> Description</th>
+</tr><tr>
+<td>
+
+``` C++
+make_platform<backend::ext_oneapi_level_zero>(
+    const backend_input_t<
+        backend::ext_oneapi_level_zero, platform> &)
+```
+</td>
+<td>Constructs a SYCL platform instance from a Level-Zero <code>ze_driver_handle_t</code>. The SYCL execution environment contains a fixed number of platforms that are enumerated via <code>sycl::platform::get_platforms()</code>. Calling this function does not create a new platform. Rather it merely creates a <code>sycl::platform</code> object that is a copy of one of the platforms from that enumeration.</td>
+</tr><tr>
+<td>
+
+``` C++
+make_device<backend::ext_oneapi_level_zero>(
+    const backend_input_t<
+        backend::ext_oneapi_level_zero, device> &)
+```
+</td>
+<td>Constructs a SYCL device instance from a Level-Zero <code>ze_device_handle_t</code>. The SYCL execution environment contains a fixed number of devices that are enumerated via <code>sycl::device::get_devices()</code> and <code>sycl::device::get_sub_devices(...)</code>. Calling this function does not create a new device. Rather it merely creates a <code>sycl::device</code> object that is a copy of one of the devices from those enumerations.</td>

```

Comment:
I'm not sure the language you added about sub-devices here is correct.  Note that the API is named `device::create_sub_devices()` (not `get_sub_devices()`).  The distinction is important because it implies that the function creates a new object, it doesn't return some pre-existing object.

The core SYCL spec does **not** say that the number of sub-devices in the execution environment is fixed, and it does **not** say that calls to `create_sub_devices()` return a copy of some existing device object.  In fact, I would expect that on many implementations that two calls to `create_sub_devices()` would typically return distinct `device` objects.

I do see that the [Level Zero documentation](https://spec.oneapi.io/level-zero/latest/core/api.html#zedevicegetsubdevices) says:

> Multiple calls to this function will return identical device handles, in the same order.

However, this is not a guarantee that the upper layers of the DPC++ runtime guarantee that multiple calls to `device::create_sub_devices()` return the same `device` object.  Therefore, it seems safer to me **not** to make this guarantee in the Level Zero interop API.

Do you have some particular reason why we should make this guarantee for sub-devices?

Reply:
I think we should make this guarantee with Level-Zero backend, since we **can** do it. It is also intuitively expected for interop with Level-Zero, which makes such guarantees as you have noted.

Reply:
After looking at the implementation, I understand this now.  Level Zero only supports partition by affinity domain, and in that case it does make sense that the set of sub-devices can be fixed.  If we want to make this guarantee, I'd like to tweak the wording to say:

> The SYCL execution environment for the Level Zero backend contains a fixed number of devices that are enumerated via <code>sycl::device::get_devices()</code> and a fixed number of sub-devices that are enumerated via <code>sycl::device::create_sub_devices(...)</code>.  [...]

I still wonder if we are painting ourselves into a corner here, though.  What if Level Zero supports `partition_by_counts` in the future?  There can not be a fixed set of sub-devices in that case, so we won't be able to support this guarantee.

Reply:
I've updated wording per your suggestion. Different partitioning is not planned in Level-Zero, and if it ever changes we would update our interop support.